service/backend/pchain/indexer/parser.go

```diff
@@ -0,0 +1,318 @@
+	}
+
+	return p.pChainClient.GetHeight(ctx)
+}
+
+func (p *parser) Initialize(ctx context.Context) (*ParsedGenesisBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	errs := wrappers.Errs{}
+
+	bytes, avaxAssetID, err := genesis.FromConfig(genesis.GetConfig(p.networkID))
+	errs.Add(err)
+	p.avaxAssetID = avaxAssetID
+
+	genesisState, err := pGenesis.Parse(bytes)
+	errs.Add(err)
+
+	blockTime := new(time.Time)
+	p.genesisTimestamp = time.Unix(int64(genesisState.Timestamp), 0)
+	*blockTime = p.genesisTimestamp
+
+	var genesisTxs []*txs.Tx
+	genesisTxs = append(genesisTxs, genesisState.Validators...)
+	genesisTxs = append(genesisTxs, genesisState.Chains...)
+
+	// Genesis commit block's parent ID is the hash of genesis state
+	var genesisParentID ids.ID = hashing.ComputeHash256Array(bytes)
+
+	// Genesis Block is not indexed by the indexer, but its block ID can be accessed from block 0's parent id
+	genesisChildBlock, err := p.ParseBlockAtIndex(ctx, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, utxo := range genesisState.UTXOs {
+		utxo.UTXO.Out.InitCtx(p.ctx)
+	}
+
+	genesisBlockID := genesisChildBlock.ParentID
+
+	return &ParsedGenesisBlock{
+		ParsedBlock: ParsedBlock{
+			ParentID:  genesisParentID,
+			Height:    0,
+			BlockID:   genesisBlockID,
+			BlockType: "GenesisBlock",
+			Timestamp: blockTime.UnixMilli(),
+			Txs:       genesisTxs,
+			Proposer:  Proposer{},
+		},
+		GenesisBlockData: GenesisBlockData{
+			Message:       genesisState.Message,
+			InitialSupply: genesisState.InitialSupply,
+			UTXOs:         genesisState.UTXOs,
+		},
+	}, errs.Err
+}
+
+func (p *parser) ParseCurrentBlock(ctx context.Context) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	height, err := p.GetPlatformHeight(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.ParseBlockAtIndex(ctx, height)
+}
+
+func (p *parser) ParseBlockAtIndex(ctx context.Context, index uint64) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByIndex(ctx, index-1)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) ParseBlockWithHash(ctx context.Context, hash string) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	hashID, err := ids.FromString(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByID(ctx, hashID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) parseBlockBytes(proposerBytes []byte) (*ParsedBlock, error) {
+	errs := wrappers.Errs{}
+
+	proposer, bytes, err := getProposerFromBytes(proposerBytes)
+	if err != nil {
+		return nil, fmt.Errorf("fetching proposer from block bytes errored with %w", err)
+	}
+
+	if p.genesisTimestamp.IsZero() {
+		return nil, errParserUninitialized
+	}
```

Comment:
nit: move it before pulling proposer up, for an early return in case parser it's not duly initialized